File: Python/euler014.py
def memoize(f):
    results = {}
    def helper(n):
        if n not in results:
            results[n] = f(n)
        return results[n]
    return helper

def main():
    maxi = 0
    lenMaxi = 0
    for i in range(2, 1000001):
        if getLength(i) > lenMaxi:
            lenMaxi = getLength(i)
            maxi = i
    return str(maxi)

@memoize
def getLength(num):
    if num == 1:
        return 1
    if num % 2 == 0:
        nextNum = num // 2
    else:
        nextNum = num * 3 + 1
    
    return getLength(nextNum) + 1

if __name__ == "__main__":
    print(main())

File: GO/euler014.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(solve())
}

func solve() int {
	maxi := 0
	lenMaxi := 0
	temp := 0
	m := make(map[int]int)

	for i := 2; i < 1000001; i++ {
		temp = get_length(i, m)
		m[i] = temp
		if temp > lenMaxi {
			lenMaxi = temp
			maxi = i
		}
	}
	return maxi
}

func get_length(n int, m map[int]int) int {
	var nextNum int
	value, ok := m[n]
	if ok {
		return value
	}
	if n == 1 {
		return 1
	}
	if n%2 == 0 {
		nextNum = int(n / 2)
	} else {
		nextNum = n*3 + 1
	}
	return get_length(nextNum, m) + 1
}
